js/noms/src/value-store.js

```diff
@@ -193,24 +201,34 @@ class SizeCache<T> {
     return entry ? entry.value : undefined;
   }
 
-  add(hash: Hash, size: number, value: ?T): void {
+  add(hash: Hash, size: Promise<number>, value: T): void {
     const key = hash.toString();
-    if (this._cache.has(key)) {
+    let entry = this._cache.get(key);
+    if (entry) {
       this._cache.delete(key);
-    } else {
-      this._size += size;
+      this._size -= entry.size;
```

Comment:
fixed. now it's totally FUNCTIONAL. expire() now waits until it has all the sizes and the boots people out.